monai/apps/deepgrow/transforms.py

```diff
@@ -0,0 +1,459 @@
+    """
+    Add random guidance as initial seed point for a given label.
+
+    Note that the label is of size (C, D, H, W) or (C, H, W)
+    The guidance is of size (2, N, # of dims) where N is number of guidance added
+    # of dims = 4 when C, D, H, W; # of dims = 3 when (C, H, W)
+
+    Args:
+        label: label source.
+        guidance: key to store guidance.
+        sids: key that represents list of valid slice indices for the given label.
+        sid: key that represents the slice to add initial seed point.  If not present, random sid will be chosen.
+        connected_regions: maximum connected regions to use for adding initial points.
+    """
+
+    def __init__(
+        self,
+        label: str = "label",
+        guidance: str = "guidance",
+        sids: str = "sids",
+        sid: str = "sid",
+        connected_regions: int = 5,
+    ):
+        self.label = label
+        self.sids = sids
+        self.sid = sid
+        self.guidance = guidance
+        self.connected_regions = connected_regions
+
+    def randomize(self, data=None):
+        pass
+
+    def _apply(self, label, sid):
+        dimensions = 3 if len(label.shape) > 3 else 2
+        default_guidance = [-1] * (dimensions + 1)
+
+        dims = dimensions
+        if sid is not None and dimensions == 3:
+            dims = 2
+            label = label[0][sid][np.newaxis]  # Assume channel is first
+
+        label = (label > 0.5).astype(np.float32)
+        blobs_labels = measure.label(label.astype(int), background=0) if dims == 2 else label
+        assert np.max(blobs_labels) > 0, "Not a valid Label"
+
+        pos_guidance = []
+        for ridx in range(1, 2 if dims == 3 else self.connected_regions + 1):
+            if dims == 2:
+                label = (blobs_labels == ridx).astype(np.float32)
+                if np.sum(label) == 0:
+                    pos_guidance.append(default_guidance)
+                    continue
+
+            distance = distance_transform_cdt(label).flatten()
+            probability = np.exp(distance) - 1.0
+
+            idx = np.where(label.flatten() > 0)[0]
+            seed = self.R.choice(idx, size=1, p=probability[idx] / np.sum(probability[idx]))
+            dst = distance[seed]
+
+            g = np.asarray(np.unravel_index(seed, label.shape)).transpose().tolist()[0]
+            g[0] = dst[0]  # for debug
+            if dimensions == 2 or dims == 3:
+                pos_guidance.append(g)
+            else:
+                pos_guidance.append([g[0], sid, g[-2], g[-1]])
+
+        return np.asarray([pos_guidance, [default_guidance] * len(pos_guidance)])
+
+    def __call__(self, data):
+        sid = data.get(self.sid)
+        sids = data.get(self.sids)
+        if sids is not None:
+            if sid is None or sid not in sids:
+                sid = self.R.choice(sids, replace=False)
+        else:
+            sid = None
+        data[self.guidance] = self._apply(data[self.label], sid)
```

Comment:
I think a fresh copy of `data` should be made, so `data = dict(data)` above this line.